src/include/catalog/catalog_content.h

```diff
@@ -50,38 +50,47 @@ class CatalogContent {
     inline common::table_id_t assignNextTableID() { return nextTableID++; }
-    inline uint64_t getTableCount() const { return tableSchemas.size(); }
+    inline uint64_t getNumTables() const { return tables->getEntries().size(); }
 
     bool containsTable(const std::string& tableName) const;
-    bool containsTable(common::TableType tableType) const;
+    bool containsTable(CatalogEntryType catalogType) const;
 
     std::string getTableName(common::table_id_t tableID) const;
 
-    TableSchema* getTableSchema(common::table_id_t tableID) const;
-    std::vector<TableSchema*> getTableSchemas(common::TableType tableType) const;
+    CatalogEntry* getTableCatalogEntry(common::table_id_t tableID) const;
+
+    template<typename T>
+    std::vector<T> getTableCatalogEntries(CatalogEntryType catalogType) const {
+        std::vector<T> result;
+        for (auto& [_, entry] : tables->getEntries()) {
+            if (entry->getType() == catalogType) {
+                result.push_back(common::ku_dynamic_cast<CatalogEntry*, T>(entry.get()));
+            }
+        }
+        return result;
+    }
 
     common::table_id_t getTableID(const std::string& tableName) const;
-    std::vector<common::table_id_t> getTableIDs(common::TableType tableType) const;
+    std::vector<common::table_id_t> getTableIDs(CatalogEntryType catalogType) const;
 
-    common::table_id_t addNodeTableSchema(const binder::BoundCreateTableInfo& info);
-    common::table_id_t addRelTableSchema(const binder::BoundCreateTableInfo& info);
-    common::table_id_t addRelTableGroupSchema(const binder::BoundCreateTableInfo& info);
-    common::table_id_t addRdfGraphSchema(const binder::BoundCreateTableInfo& info);
-    void dropTableSchema(common::table_id_t tableID);
+    common::table_id_t createNodeTable(const binder::BoundCreateTableInfo& info);
+    common::table_id_t createRelTable(const binder::BoundCreateTableInfo& info);
+    common::table_id_t createRelGroup(const binder::BoundCreateTableInfo& info);
+    common::table_id_t createRDFGraph(const binder::BoundCreateTableInfo& info);
+    void dropTable(common::table_id_t tableID);
     void renameTable(common::table_id_t tableID, const std::string& newName);
 
 private:
-    // TODO(Guodong): I don't think it's necessary to keep separate maps for node and rel tables.
-    std::unordered_map<common::table_id_t, std::unique_ptr<TableSchema>> tableSchemas;
     // These two maps are maintained as caches. They are not serialized to the catalog file, but
     // is re-constructed when reading from the catalog file.
     std::unordered_map<std::string, common::table_id_t> tableNameToIDMap;
     common::table_id_t nextTableID;
     std::unique_ptr<function::BuiltInFunctions> builtInFunctions;
     std::unordered_map<std::string, std::unique_ptr<function::ScalarMacroFunction>> macros;
     common::VirtualFileSystem* vfs;
+    std::unique_ptr<CatalogSet> tables;
```

Comment:
If we introduce CatalogSet here (btw you probably don't need unique_ptr), I have following questions

- should we remove `std::unordered_map<std::string, common::table_id_t> tableNameToIDMap;` and `nextTableID` ?
- Do we have plan to make `macros` and `builtInFunctions` as part of CatalogSet

Reply:
macros and builtInFunctions will have their own catalogSet. 
The schema catalogEntry would contain the following fields (copied from duckdb):
```
class DuckSchemaEntry : public SchemaCatalogEntry {
public:
	DuckSchemaEntry(Catalog &catalog, string name, bool is_internal);

private:
	//! The catalog set holding the tables
	CatalogSet tables;
	//! The catalog set holding the indexes
	CatalogSet indexes;
	//! The catalog set holding the table functions
	CatalogSet table_functions;
	//! The catalog set holding the copy functions
	CatalogSet copy_functions;
	//! The catalog set holding the pragma functions
	CatalogSet pragma_functions;
	//! The catalog set holding the scalar and aggregate functions
	CatalogSet functions;
	//! The catalog set holding the sequences
	CatalogSet sequences;
	//! The catalog set holding the collations
	CatalogSet collations;
	//! The catalog set holding the types
	CatalogSet types;
}
```